cloud/scope/azureippool.go

```diff
@@ -0,0 +1,348 @@
+		machineScope.Info(fmt.Sprintf("Setting ReconciledIP to false"))
+		machineScope.AzureMachine.Status.ReconciledIP = false
+
+		//patch azure machine
+		if err := machineScope.patchHelper.Patch(ctx, machineScope.AzureMachine); err != nil {
+			return errors.Wrapf(err, "Failed to update Azure Machine")
+		}
+		machineScope.Info(fmt.Sprintf("Successfully updated AzureMachine status"))	
+		
+	}				
+
+	return nil
+}
+
+func BuildNetworkInterfaceSpec(machineScope *MachineScope,ippool infrav1.IPPool, name string) infrav1.NetworkInterface {
+    networkInterface := infrav1.NetworkInterface{}
+    networkInterface.Name = azure.GenerateNICName(machineScope.Name())
+    networkInterface.VnetName = ippool.VnetName
+    networkInterface.SubnetName = ippool.SubnetName
+    networkInterface.VnetResourceGroup = "aserg"
+    if name == "mgmt-nic-ip-pool" {
+        networkInterface.IsPrimary = true
+    } else {
+		networkInterface.AcceleratedNetworking = true
+    }
+    return networkInterface
+}
+
+func GetPrimaryNetworkInterface(nics []infrav1.NetworkInterface) (infrav1.NetworkInterface, error) {
+    nic := infrav1.NetworkInterface{}
+	for _,nic := range(nics) {
+        if nic.IsPrimary == true {
+            return nic, nil
+        }
+    }
+    return nic, errors.New("Unable to find a nic with isPrimary set to true")
+}
+
+func AddToAllocatedIPPool(azureippool *infrav1.AzureIPPool, name string, allocatedIP string, vnetName string, subnetName string) *infrav1.AzureIPPool {
+	
+	//iterate through all the allocated IP pools to find if allocated IP pool exists
+	for index, allocatedIPPool := range(azureippool.Status.AllocatedIPPools) {
+		if allocatedIPPool.Name == name {
+			allocatedIPPool.AllocatedIPs = append(allocatedIPPool.AllocatedIPs, allocatedIP)
+			azureippool.Status.AllocatedIPPools[index] = allocatedIPPool
+			return azureippool
+		}
+	}
+
+	allocatedIPPool := InitializeAllocatedIPPool(name, allocatedIP, vnetName, subnetName)
+	azureippool.Status.AllocatedIPPools = append(azureippool.Status.AllocatedIPPools, allocatedIPPool)
+	return azureippool	
+}
+
+func RemoveIPFromAllocatedIPPool(allocatedIPPool infrav1.AllocatedIPPool, freeIP string) (infrav1.AllocatedIPPool, bool) {
+	log := klogr.New()
+	for i, ip := range(allocatedIPPool.AllocatedIPs) {
+		if ip == freeIP {
+			log.Info(fmt.Sprintf("Found IP in allocated ip pool: %s", freeIP))
+			allocatedIPPool.AllocatedIPs = append(allocatedIPPool.AllocatedIPs[0:i],allocatedIPPool.AllocatedIPs[i+1:]...) 
+			return allocatedIPPool, true
+		}
+	}
+	return allocatedIPPool, false
+}
+
+func InitializeAllocatedIPPool(name string, allocatedIP string, vnetName string, subnetName string) infrav1.AllocatedIPPool {
+	allocatedIPPool := infrav1.AllocatedIPPool {
+		Name : name,
+		AllocatedIPs : []string{allocatedIP},
+		VnetName : vnetName,
+		SubnetName : subnetName,
+	}
+	return allocatedIPPool
+}
+
+func GetNextFreeIP(ip string, IPs []string, allocatedIPsMap map[string]bool) (string, error) {
```

Comment:
Nit. I think this can be split into 2 different functions:
GetNextFreeIP(...)  <-- find the next free ip address
AllocateIP(ip) <-- only try and allocate the given ip address.

Reply:
@pradipd NextFreeIP just returns the next free IP- API server IP for control plane VM and any available free IP for worker VM. There is another function AddToAllocatedIPPool that updates this IP on the machine status.